src/shell/autoexec.cpp

```diff
@@ -0,0 +1,675 @@
+
+	const auto blob_bin = std::vector<uint8_t>(blob_dos.begin(), blob_dos.end());
+	if (is_blob_registered) {
+		VFILE_Update(autoexec_file_name.c_str(), blob_bin);
+	} else {
+		VFILE_Register(autoexec_file_name.c_str(), blob_bin);
+		is_blob_registered = true;
+	}
+}
+
+// ***************************************************************************
+// AutoExecObject class implementation
+// ***************************************************************************
+
+void AutoExecObject::SetString(const std::string& content, const Source source)
+{
+	assert(source != Source::Variable); // use SetVariable for this source
+
+	ReInstall(autoexec_objects[source]);
+
+	string_content = content;
+}
+
+void AutoExecObject::SetVariable(const std::string& name, const std::string& value)
+{
+#if C_DEBUG
+	auto assert_printable_ascii = [](const std::string& in) {
+		for (const auto character : in) {
+			if (!is_printable_ascii(character)) {
+				E_Exit("AUTOEXEC: Not a printable ASCII");
+			}
+		}
+	};
+
+	assert_printable_ascii(name);
+	assert_printable_ascii(value);
+#endif
+	if (name == variable_name && value == variable_value) {
+		// Same variable, same value - nothing to do
+		return;
+	}
+
+	ReInstall(autoexec_objects[Source::Variable]);
+
+	const auto old_name = variable_name;
+
+	variable_name  = name;
+	variable_value = value;
+	upcase(variable_name);
+
+	// If shell already exists, refresh variable content
+	if (first_shell) {
+		if (!old_name.empty() && variable_name != old_name) {
+			first_shell->SetEnv(old_name.c_str(), "");
+		}
+		first_shell->SetEnv(variable_name.c_str(), variable_value.c_str());
+	}
+}
+
+std::string AutoExecObject::GetContent() const
+{
+	// Handle environment variables
+
+	if (!variable_name.empty()) {
+		const std::string set = "@SET ";
+		return set + variable_name + "=" + variable_value;
+	}
+
+	// Handle string content
+
+	return string_content;
+}
+
+void AutoExecObject::ReInstall(std::list<AutoExecObject*>& list)
+{
+	if (std::find(list.begin(), list.end(), this) == list.end()) {
+		// Remove from other lists
+		UnInstall();
+	}
+
+	if (!is_installed) {
+		list.push_back(this);
+		is_installed = true;
+	}
+}
+
+void AutoExecObject::UnInstall()
+{
+	if (!is_installed) {
+		return;
+	}
+
+	// For environment variable, remove it from runningg shell
+	if (first_shell && !variable_name.empty()) {
+		first_shell->SetEnv(variable_name.c_str(), "");
+	}
+
+	// Remove object from all the lists
+	for (auto& [source, list] : autoexec_objects) {
+		list.remove(this);
+	}
+
+	// Clear internal data
+	string_content.clear();
+	variable_name.clear();
+	variable_value.clear();
+
+	is_installed = false;
+}
+
+AutoExecObject::~AutoExecObject()
+{
+	UnInstall();
+}
+
+// ***************************************************************************
+// AUTOEXEC class declaration and implementation
+// ***************************************************************************
+
+class AUTOEXEC final : public Module_base {
```

Comment:
`Autoexec` or `AutoExec`

Reply:
Renamed to `AutoExecModule`, to distinguish from [autoexec] section content and AUTOEXEC.BAT content.